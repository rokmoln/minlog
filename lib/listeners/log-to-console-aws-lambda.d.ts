import { MinLogListener } from '../types';
export declare let logToConsoleAwsLambda: (cfg?: {
    level?: string | number;
}) => MinLogListener;
export default logToConsoleAwsLambda;
