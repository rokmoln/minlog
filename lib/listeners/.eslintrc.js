"use strict";module.exports = {
  extends: [
  'firecloud/browser.js'] };

//# sourceMappingURL=.eslintrc.js.map