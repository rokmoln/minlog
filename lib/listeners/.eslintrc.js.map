{"version":3,"sources":["../../src/listeners/.eslintrc.js"],"names":[],"mappings":";;AAAA,MAAM,CAAC,OAAP,GAAiB;AACf,EAAA,OAAO,EAAE,CACP,sBADO;AADM,CAAjB","file":".eslintrc.js","sourcesContent":["module.exports = {\n  extends: [\n    'firecloud/browser.js'\n  ]\n};\n"]}