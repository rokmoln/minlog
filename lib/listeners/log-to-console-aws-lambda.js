"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.default = exports.logToConsoleAwsLambda = exports.format = exports._isAwsLambda = exports._nonBreakingWhitespace = void 0;var _lodashFirecloud = _interopRequireDefault(require("lodash-firecloud"));
var _fastSafeStringify = _interopRequireDefault(require("fast-safe-stringify"));







var _util = require("../util");




var _logToConsole = require("./log-to-console");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}


















let _nonBreakingWhitespace = 'Â ';exports._nonBreakingWhitespace = _nonBreakingWhitespace;

let _isAwsLambda = function () {
  let isNode = typeof process !== 'undefined' && _lodashFirecloud.default.isDefined(_lodashFirecloud.default.get(process, 'versions.node'));
  if (!isNode) {
    return false;
  }

  let isAwsLambda = _lodashFirecloud.default.isDefined(process.env.LAMBDA_TASK_ROOT);
  if (!isAwsLambda) {
    return false;
  }

  return true;
};exports._isAwsLambda = _isAwsLambda;

let format = function (...formatArgs) {
  let [
  format,
  ...args] =
  formatArgs;

  // eslint-disable-next-line @typescript-eslint/no-require-imports
  let chunk = require('util').format(format, ...args);
  chunk = _lodashFirecloud.default.replace(chunk, /\n/g, '\r');
  chunk = `${chunk}\n`;
  process.stdout.write(chunk);
};exports.format = format;

let logToConsoleAwsLambda = function (cfg = {}) {
  if (!exports._isAwsLambda()) {
    // use vanilla logger e.g. behind aws-lambda-proxy
    return (0, _logToConsole.logToConsole)(cfg);
  }

  // from https://github.com/Financial-Times/lambda-logger
  // This does make process.stdout.write a blocking function (process.stdout._handle.setBlocking(true);),
  // as AWS Lambda previously streamed to an output which was synchronous,
  // but has since changed to asynchronous behaviour, leading to lost logs.
  if (_lodashFirecloud.default.isFunction(_lodashFirecloud.default.get(process, 'stdout._handle.setBlocking'))) {
    // @ts-ignore
    process.stdout._handle.setBlocking(true);
  }

  // eslint-disable-next-line complexity
  return async function ({ entry, logger, rawEntry }) {
    cfg = _lodashFirecloud.default.isFunction(cfg) ? await (async createError => {try {return await cfg();} catch (_awaitTraceErr) {let err = createError();_awaitTraceErr.stack += "\n...\n" + err.stack;throw _awaitTraceErr;}})(() => new Error()) : await (async createError => {try {return await cfg;} catch (_awaitTraceErr2) {let err = createError();_awaitTraceErr2.stack += "\n...\n" + err.stack;throw _awaitTraceErr2;}})(() => new Error());

    if (_lodashFirecloud.default.isDefined(rawEntry) &&
    _lodashFirecloud.default.filter(rawEntry._args).length === 1 &&
    _lodashFirecloud.default.isDefined(rawEntry._args[0]._babelSrc)) {
      return;
    }

    if (_lodashFirecloud.default.isDefined(cfg.level) && logger.levelIsBeyondGroup(entry._level, cfg.level)) {
      return;
    }

    let {
      cfg: cfg2,
      now,
      formattedLevelName,
      src,
      msg
      // @ts-ignore
    } = (0, _logToConsole.serialize)({ entry, logger, rawEntry, cfg });
    cfg = cfg2;

    // prefer JSON output over util.inspect output
    let extra = _lodashFirecloud.default.omit(entry, [
    '_args']);

    let extraStr = (0, _fastSafeStringify.default)(extra, _util.jsonStringifyReplacer, 2);
    extraStr = `\n${extraStr}`;

    // maintain whitespace (looking at you AWS CloudWatch WebUI)
    // by replacing space with non-breaking space
    extraStr = _lodashFirecloud.default.replace(extraStr, / /g, exports._nonBreakingWhitespace);

    let extraArgs = [
    extraStr];


    let formatPairs = [];

    // timestamp
    formatPairs.push([
    '%s',
    now]);


    // awsRequestId
    formatPairs.push([
    '\t%s',
    _lodashFirecloud.default.get(entry, 'ctx.awsRequestId', '-')]);


    // level name
    formatPairs.push([
    '\t%s',
    formattedLevelName]);


    // src
    if (_lodashFirecloud.default.isDefined(src)) {
      formatPairs.push([
      '\t%s',
      src]);

    }

    // msg
    // eslint-disable-next-line lodash/prop-shorthand
    let argNames = _lodashFirecloud.default.keys((0, _util.keepOnlyExtra)(entry));
    if (!_lodashFirecloud.default.isEmpty(argNames)) {
      msg = `${msg} (${_lodashFirecloud.default.join(argNames, ', ')})`;
    }
    msg = `${msg} `;
    msg = _lodashFirecloud.default.padEnd(msg, 255, exports._nonBreakingWhitespace);
    msg = `${msg}.`;
    formatPairs.push([
    '\n%s',
    msg]);


    let formatArgs = (0, _logToConsole.toFormatArgs)(...formatPairs);
    formatArgs.push(...extraArgs);

    exports.format(...formatArgs);
  };
};exports.logToConsoleAwsLambda = logToConsoleAwsLambda;var _default = exports.logToConsoleAwsLambda;exports.default = _default;

//# sourceMappingURL=log-to-console-aws-lambda.js.map