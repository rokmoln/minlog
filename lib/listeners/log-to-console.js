"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.default = exports.logToConsole = exports._levelToConsoleFun = exports._isAwsLambda = exports._isNode = exports._isBrowser = void 0;var _lodashFirecloud = _interopRequireDefault(require("lodash-firecloud"));
var _fastSafeStringify = _interopRequireDefault(require("fast-safe-stringify"));
var _moment = _interopRequireDefault(require("moment"));function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}

let _isBrowser = typeof window !== 'undefined';exports._isBrowser = _isBrowser;
let _isNode = typeof process !== 'undefined' && !_lodashFirecloud.default.isUndefined(_lodashFirecloud.default.get(process, 'versions.node'));exports._isNode = _isNode;
let _isAwsLambda = exports._isNode && !_lodashFirecloud.default.isUndefined(process.env.LAMBDA_TASK_ROOT);

// from https://github.com/Financial-Times/lambda-logger
// This does make process.stdout.write a blocking function (process.stdout._handle.setBlocking(true);),
// as AWS Lambda previously streamed to an output which was synchronous,
// but has since changed to asynchronous behaviour, leading to lost logs.
exports._isAwsLambda = _isAwsLambda;if (exports._isAwsLambda && _lodashFirecloud.default.isFunction(_lodashFirecloud.default.get(process, 'stdout._handle.setBlocking'))) {
  process.stdout._handle.setBlocking(true);
}

let _levelToConsoleFun = function ({ level, levels }) {
  // on AWS Lambda the console.trace call will print '[object Object]'¯\_(ツ)_/¯
  // plus all console funs end up in a cloudwatch stream
  if (exports._isAwsLambda) {
    return 'log';
  }

  if (_lodashFirecloud.default.isString(level)) {
    // eslint-disable-next-line prefer-destructuring
    level = levels[level];
  }

  if (_lodashFirecloud.default.inRange(level, 0, levels.warn)) {
    return 'error';
  } else if (_lodashFirecloud.default.inRange(level, levels.warn, levels.info)) {
    return 'warn';
  } else if (_lodashFirecloud.default.inRange(level, levels.info, levels.debug)) {
    return 'info';
  } else if (_lodashFirecloud.default.inRange(level, levels.debug, levels.trace)) {
    // return 'debug';
    // console.debug doesn't seem to print anything,
    // but console.debug is an alias to console.log anyway
    return 'log';
  } else if (level === levels.trace) {
    return 'trace';
  }

  return 'log';
};

/*
   cfg has 2 properties
   - contextId (optional, default to 'top' in browser and undefined elsewhere)
     An identifier for the current "context".
   - level (optional, defaults to trace)
     Any log entry less important that cfg.level is ignored.
   */exports._levelToConsoleFun = _levelToConsoleFun;

let logToConsole = function (cfg = {}) {
  let contextId;
  let hasCssSupport = false;

  if (exports._isBrowser) {
    if (window.parent === window) {
      contextId = 'top';
    }
    hasCssSupport = true;
  }

  _lodashFirecloud.default.defaults(cfg, {
    contextId,
    level: 'trace' });


  let maybeCss = function (css) {
    return hasCssSupport ? css : '';
  };

  // eslint-disable-next-line complexity
  return async function ({ entry, logger, rawEntry }) {
    if (_lodashFirecloud.default.filter(rawEntry._args).length === 1 && rawEntry._args[0]._babelSrc) {
      return;
    }

    if (logger.levelIsBeyondGroup(entry._level, cfg.level)) {
      return;
    }

    let now = (0, _moment.default)(entry._time.stamp).utcOffset(entry._time.utc_offset).toISOString();
    let levelName = logger.levelToLevelName(entry._level);
    let formattedLevelName = _lodashFirecloud.default.padStart(_lodashFirecloud.default.toUpper(levelName), '5');
    let consoleFun = exports._levelToConsoleFun({
      level: entry._level,
      levels: logger.levels });


    let color = '';
    switch (consoleFun) {
      case 'log':
      case 'info':
      case 'trace':
        color = maybeCss('color: dodgerblue');
        break;
      default:}


    let nowAndContextSep = _lodashFirecloud.default.isAwsLambda ? '\t' : ' ';
    let maybeContextFormat = _lodashFirecloud.default.isUndefined(cfg.contextId) ? '' : `${nowAndContextSep}%s`;

    let src = '';
    if (entry._babelSrc) {
      src = _lodashFirecloud.default.merge({}, entry._src, entry._babelSrc);
      src = ` @webpack:///./${src.file}:${src.line}:${src.column}${src.function ? ` in ${src.function}()` : ''}`;
    } else if (entry._src) {
      src = entry._src;
      src = ` ${src.file}:${src.line}:${src.column}${src.function ? ` in ${src.function}()` : ''}`;
    }

    let prefixFormat = `%c%s${maybeContextFormat}${nowAndContextSep}%c%s%c%s`;
    if (!hasCssSupport) {
      prefixFormat = _lodashFirecloud.default.replace(prefixFormat, /%c/g, hasCssSupport ? '%c' : '%s');
    }
    let prefixArgs = [
    maybeCss(color),
    now,
    _lodashFirecloud.default.isUndefined(cfg.contextId) ? '' : cfg.contextId,
    maybeCss('font-weight: bold'),
    formattedLevelName,
    maybeCss(color),
    src];


    let msgFormat = '';
    let msgArgs = [];
    if (entry.msg) {
      let {
        _duration: duration,
        msg } =
      entry;
      if (duration) {
        msg = `${msg} - took ${duration.ms} ms`;
        if (duration.ms > 1000) {
          msg = `${msg} (${duration.human})`;
        }
      }

      msgFormat = '\n%s';
      msgArgs = [
      msg];

    }

    let extraFormat = '';
    let extraArgs = [];

    let extra = _lodashFirecloud.default.omit(entry, [
    '_args',
    '_babelSrc',
    '_duration',
    '_level',
    '_src',
    '_time',
    '_timeEnd',
    '_timeStart',
    'contextId',
    'msg']);

    extra = _lodashFirecloud.default.omitBy(extra, function (_value, key) {
      return /^_arg[0-9+]$/.test(key);
    });

    let context = {};
    if (exports._isBrowser) {
      _lodashFirecloud.default.merge(context, {
        window,
        documentElement: window.document.documentElement });

    }
    _lodashFirecloud.default.merge(extra, context);

    // devTools console sorts keys when object is expanded
    extra = _lodashFirecloud.default.toPairs(extra);
    extra = _lodashFirecloud.default.sortBy(extra, 0);
    extra = _lodashFirecloud.default.fromPairs(extra);

    // devTools collapses objects with 'too many' keys,
    // so we output objects with only one key
    _lodashFirecloud.default.forEach(extra, function (value, key) {
      extraArgs.push('\n');
      let obj = {
        [key]: value };


      // fix for util.inspect having no indentation
      if (!exports._isBrowser) {
        obj = (0, _fastSafeStringify.default)(obj, undefined, 2);
      }

      extraArgs.push(obj);
    });

    let format = `${prefixFormat}:${msgFormat}${extraFormat}`;
    let vars = [
    ...prefixArgs,
    ...msgArgs,
    ...extraArgs];


    if (exports._isAwsLambda) {
      // eslint-disable-next-line global-require
      let msg = require('util').format(format, ...vars);
      msg = _lodashFirecloud.default.replace(msg, /\n/g, '\r');
      msg = `${msg}\n`;
      process.stdout.write(msg);
      return;
    }

    // eslint-disable-next-line no-console
    console[consoleFun](format, ...vars);
  };
};exports.logToConsole = logToConsole;var _default = exports.logToConsole;exports.default = _default;

//# sourceMappingURL=log-to-console.js.map