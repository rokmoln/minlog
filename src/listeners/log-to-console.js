import _ from 'lodash-firecloud';
import fastSafeStringify from 'fast-safe-stringify';

let _isBrowser = typeof window !== 'undefined';
let _isNode = typeof process !== 'undefined' && _.isDefined(_.get(process, 'versions.node'));

let _levelToConsoleFun = function({level, levels}) {
  if (_.isString(level)) {
    // eslint-disable-next-line prefer-destructuring
    level = levels[level];
  }

  if (_.inRange(level, 0, levels.warn)) {
    return 'error';
  } else if (_.inRange(level, levels.warn, levels.info)) {
    return 'warn';
  } else if (_.inRange(level, levels.info, levels.debug)) {
    return 'info';
  } else if (_.inRange(level, levels.debug, levels.trace)) {
    // return 'debug';
    // console.debug doesn't seem to print anything,
    // but console.debug is an alias to console.log anyway
    return 'log';
  } else if (level === levels.trace) {
    return 'trace';
  }

  return 'log';
};

/*
cfg has 2 properties
- contextId (optional, default to 'top' in browser and undefined elsewhere)
  An identifier for the current "context".
- level (optional, defaults to trace)
  Any log entry less important that cfg.level is ignored.
*/

export let serialize = function({entry, logger, _rawEntry, cfg}) {
  let contextId;
  let hasCssSupport = false;

  if (_isBrowser) {
    if (window.parent === window) {
      contextId = 'top';
    }
    hasCssSupport = true;
  }

  cfg = _.defaults({}, cfg, {
    contextId,
    level: 'trace',
    localTime: false
  });

  let now = cfg.localTime ?
    entry._time.localStamp :
    entry._time.stamp;
  let levelName = logger.levelToLevelName(entry._level);
  let formattedLevelName = _.padStart(_.toUpper(levelName), 5);
  let consoleFun = _levelToConsoleFun({
    level: entry._level,
    levels: logger.levels
  });

  let color = '';
  switch (consoleFun) {
  case 'log':
  case 'info':
  case 'trace':
    color = 'color: dodgerblue';
    break;
  default:
  }

  let src = _.merge({}, entry._src, entry._babelSrc);
  if (_.isEmpty(src)) {
    src = '';
  } else if (_isBrowser) {
    // FIXME assumes webpack; could be a cfg flag, unless webpack can be detected
    src = `@webpack:///./${src.file}:${src.line}:${src.column}${src.function ? ` in ${src.function}()` : ''}`;
  } else {
    src = `${src.file}:${src.line}:${src.column}${src.function ? ` in ${src.function}()` : ''}`;
  }

  let msg = entry.msg || '';
  if (msg) {
    let {
      _duration: duration
    } = entry;

    if (duration) {
      msg = `${msg} - took ${duration.ms} ms`;
      if (duration.ms > 1000) {
        msg = `${msg} (${duration.human})`;
      }
    }
  }

  let extra = _.omit(entry, [
    '_args',
    '_babelSrc',
    '_duration',
    '_level',
    '_src',
    '_time',
    '_timeEnd',
    '_timeStart',
    'msg'
  ]);

  if (_isBrowser) {
    let ctx = {
      window,
      documentElement: window.document.documentElement
    };
    _.merge(extra, {
      ctx
    });
  }

  // devTools console sorts keys when object is expanded
  extra = _.toPairs(extra);
  extra = _.sortBy(extra, 0);
  extra = _.fromPairs(extra);

  return {
    cfg,
    hasCssSupport,

    now,
    levelName,
    formattedLevelName,
    consoleFun,
    color,
    src,
    msg,

    extra
  };
};

export let toFormatArgs = function(...args) {
  let {
    format,
    formatArgs
  } = _.reduce(args, function(result, arg) {
    let {
      format,
      formatArgs
    } = result;

    if (_.isString(arg)) {
      format.push(arg);
      return {
        format,
        formatArgs
      };
    }

    format.push(arg[0]);
    formatArgs.push(arg[1]);

    return {
      format,
      formatArgs
    };
  }, {
    format: [],
    formatArgs: []
  });

  format = _.join(format, '');
  formatArgs.unshift(format);
  return formatArgs;
};

export let logToConsole = function(cfg = {}) {
  // eslint-disable-next-line complexity
  return async function({entry, logger, rawEntry}) {
    if (_.isDefined(rawEntry) &&
        _.filter(rawEntry._args).length === 1 &&
        rawEntry._args[0]._babelSrc) {
      return;
    }

    if (logger.levelIsBeyondGroup(entry._level, cfg.level)) {
      return;
    }

    let {
      cfg: cfg2,
      hasCssSupport,
      now,
      formattedLevelName,
      consoleFun,
      color,
      src,
      msg,
      extra
    } = serialize({entry, logger, rawEntry, cfg});
    cfg = cfg2;

    let extraArgs = [];
    // devTools collapses objects with 'too many' keys,
    // so we output objects with only one key
    _.forEach(extra, function(value, key) {
      if (_.isUndefined(value)) {
        return;
      }

      value = {
        [key]: value
      };

      if (_isBrowser) {
        extraArgs.push('\n');
        extraArgs.push(value);
      } else if (_isNode) {
        // prefer JSON output over util.inspect output
        value = fastSafeStringify(value, undefined, 2);
        value = `\n${value}`;
        extraArgs.push(value);
      }
    });

    let formatArgs = [];

    // color
    if (hasCssSupport) {
      formatArgs.push([
        '%c',
        color
      ]);
    }

    // timestamp
    formatArgs.push([
      '%s',
      now
    ]);

    // context
    if (_.isDefined(cfg.contextId)) {
      formatArgs.push([
        ' %s',
        cfg.contextId
      ]);
    }

    // level name
    if (hasCssSupport) {
      formatArgs.push([
        '%c',
        'font-weight: bold'
      ]);
    }
    formatArgs.push([
      ' %s',
      formattedLevelName
    ]);

    // color
    if (hasCssSupport) {
      formatArgs.push([
        '%c',
        color
      ]);
    }

    // src
    if (src) {
      formatArgs.push([
        ' %s',
        src
      ]);
    }

    // msg
    formatArgs.push([
      '\n%s',
      msg
    ]);

    formatArgs = toFormatArgs(...formatArgs);
    formatArgs = _.concat(formatArgs, extraArgs);

    // eslint-disable-next-line no-console
    console[consoleFun](...formatArgs);
  };
};

export default logToConsole;
