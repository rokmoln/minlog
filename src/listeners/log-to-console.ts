import TypescriptMinLog from '../minlog';
import _ from 'lodash-firecloud';
import fastSafeStringify from 'fast-safe-stringify';

import {
  MinLogEntry,
  MinLogFormatArgs,
  MinLogLevel,
  MinLogLevelNameToCode,
  MinLogListener,
  MinLogRawEntry,
  MinLogSerializedTime
} from '../types';

import {
  keepOnlyExtra
} from '../util';

type FormatPair = string | [string, any];

let _isBrowser = typeof window !== 'undefined';
let _isNode = typeof process !== 'undefined' && _.isDefined(_.get(process, 'versions.node'));

let _levelToConsoleFun = function({level, levels}: {
  level: MinLogLevel,
  levels: MinLogLevelNameToCode
}): string {
  if (_.isString(level)) {
    level = levels[level];
  }

  if (_.inRange(level, 0, levels.warn)) {
    return 'error';
  } else if (_.inRange(level, levels.warn, levels.info)) {
    return 'warn';
  } else if (_.inRange(level, levels.info, levels.debug)) {
    return 'info';
  } else if (_.inRange(level, levels.debug, levels.trace)) {
    // return 'debug';
    // console.debug doesn't seem to print anything,
    // but console.debug is an alias to console.log anyway
    return 'log';
  } else if (level === levels.trace) {
    return 'trace';
  }

  return 'log';
};

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export let serialize = function({entry, logger, rawEntry: _rawEntry, cfg}: {
  entry: MinLogEntry & {
    _time: MinLogSerializedTime
  },
  logger: TypescriptMinLog,
  rawEntry: MinLogRawEntry,
  cfg?: {
    contextId?: string,
    level?: MinLogLevel,
    localTime?: boolean,
    localStamp?: number,
    stamp?: number
  }
}) {
  let contextId;
  let hasCssSupport = false;

  if (_isBrowser) {
    contextId = window.parent === window ? 'top' : 'iframe';
    hasCssSupport = true;
  }

  cfg = _.defaults({}, cfg, {
    contextId,
    level: 'trace',
    localTime: false
  });

  let now = cfg.localTime ?
    entry._time.localStamp :
    entry._time.stamp;
  let levelName = logger.levelToLevelName(entry._level);
  let formattedLevelName = _.padStart(_.toUpper(levelName), 5);
  let consoleFun = _levelToConsoleFun({
    level: entry._level,
    levels: logger.levels
  });

  let color = '';
  switch (consoleFun) {
  case 'log':
  case 'info':
  case 'trace':
    color = 'color: dodgerblue';
    break;
  default:
  }

  let src = _.merge({}, entry._src, entry._babelSrc);
  let srcStr: string;

  if (_.isEmpty(src)) {
    srcStr = '';
  } else {
    _.defaults(src, {
      // column may only be available in _babelSrc
      // prefer a default, over not printing `:?` for structured parsing, with no optional groups
      column: '?'
    });
    let inSrcFunction = _.isDefined(src.function) ? ` in ${src.function}()` : '';
    srcStr = `${src.file}:${src.file}:${src.column}${inSrcFunction}`;
  }

  let msg = _.defaultTo(entry.msg, '');
  if (!_.isEmpty(msg)) {
    let {
      _duration: duration
    } = entry;

    if (_.isDefined(duration)) {
      msg = `${msg} - took ${duration.ms} ms`;
      if (duration.ms > 1000) {
        msg = `${msg} (${duration.human})`;
      }
    }
  }

  let extra = keepOnlyExtra(entry);

  if (_isBrowser) {
    let ctx = {
      window,
      documentElement: window.document.documentElement
    };
    _.merge(extra, {
      ctx
    });
  }

  // devTools console sorts keys when object is expanded
  let extraPairs = _.toPairs(extra);
  // eslint-disable-next-line lodash/prop-shorthand
  extraPairs = _.sortBy(extraPairs, 0);
  extra = _.fromPairs(extraPairs);

  return {
    cfg,
    hasCssSupport,

    now,
    levelName,
    formattedLevelName,
    consoleFun,
    color,
    src: srcStr,
    msg,

    extra
  };
};

export let toFormatArgs = function(...formatPairs: FormatPair[]): MinLogFormatArgs {
  let {
    format,
    formatArgs
  } = _.reduce(formatPairs, function(result, formatPair) {
    let {
      format,
      formatArgs
    } = result;

    if (_.isString(formatPair)) {
      format.push(formatPair);
      return {
        format,
        formatArgs
      };
    }

    format.push(formatPair[0]);
    formatArgs.push(formatPair[1]);

    return {
      format,
      formatArgs
    };
  }, {
    format: [],
    formatArgs: []
  });

  let formatStr = _.join(format, '');
  formatArgs.unshift(formatStr);
  return formatArgs as [string, ...any[]];
};

export let format = function(consoleFun: string, ...formatArgs: MinLogFormatArgs): void {
  let [
    format,
    ...args
  ] = formatArgs;

  // eslint-disable-next-line no-console
  console[consoleFun](format, ...args);
};

export let logToConsole = function(cfg: {

  /**
   * Any log entry less important that cfg.level is ignored.
   */
  level?: MinLogLevel,

  /**
   * A context id. In a browser environment, it defaults to 'top'/'iframe'.
   */
  contextId?: string
} = {}): MinLogListener {
  // eslint-disable-next-line complexity
  return async function({entry, logger, rawEntry}) {
    // eslint-disable-next-line require-atomic-updates
    cfg = _.isFunction(cfg) ? await cfg() : await cfg;

    if (_.isDefined(rawEntry) &&
        _.filter(rawEntry._args).length === 1 &&
        _.isDefined(rawEntry._args[0]._babelSrc)) {
      return;
    }

    if (_.isDefined(cfg.level) && logger.levelIsBeyondGroup(entry._level, cfg.level)) {
      return;
    }

    let {
      cfg: cfg2,
      hasCssSupport,
      now,
      formattedLevelName,
      consoleFun,
      color,
      src,
      msg,
      extra
    // @ts-ignore
    } = serialize({entry, logger, rawEntry, cfg});
    // eslint-disable-next-line require-atomic-updates
    cfg = cfg2;

    let extraArgs = [];
    // devTools collapses objects with 'too many' keys,
    // so we output objects with only one key
    _.forEach(extra, function(value, key) {
      if (_.isUndefined(value)) {
        return;
      }

      let valueObj = {
        [key]: value
      };

      if (_isBrowser) {
        extraArgs.push('\n');
        extraArgs.push(valueObj);
      } else if (_isNode) {
        // prefer JSON output over util.inspect output
        let valueStr = fastSafeStringify(valueObj, undefined, 2);
        valueStr = `\n${valueStr}`;
        extraArgs.push(valueStr);
      }
    });

    let formatPairs = [];

    // color
    if (hasCssSupport) {
      formatPairs.push([
        '%c',
        color
      ]);
    }

    // timestamp
    formatPairs.push([
      '%s',
      now
    ]);

    // context
    if (_.isDefined(cfg.contextId)) {
      formatPairs.push([
        ' %s',
        cfg.contextId
      ]);
    }

    // level name
    if (hasCssSupport) {
      formatPairs.push([
        '%c',
        'font-weight: bold'
      ]);
    }
    formatPairs.push([
      ' %s',
      formattedLevelName
    ]);

    // color
    if (hasCssSupport) {
      formatPairs.push([
        '%c',
        color
      ]);
    }

    // src
    if (_.isDefined(src)) {
      formatPairs.push([
        ' %s',
        src
      ]);
    }

    // msg
    formatArgs.push([
      '\n%s',
      msg
    ]);

    let formatArgs = toFormatArgs(...formatPairs);
    formatArgs.push(...extraArgs);

    // eslint-disable-next-line no-console
    format(consoleFun, ...formatArgs);
  };
};

export default logToConsole;
