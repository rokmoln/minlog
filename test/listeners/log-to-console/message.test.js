import * as logToConsoleModule from '../../../src/listeners/log-to-console';
import _ from 'lodash-firecloud';

import {
  logger
} from './logger';

describe('logToConsole listener', function() {
  it('should print a message', async function() {
    let spyFormat = jest.spyOn(logToConsoleModule, 'format');

    let loggerCallArgs = [
      'This is a message.'
    ];

    spyFormat.mockImplementationOnce(function(consoleFun, format, ...formatArgs) {
      let [
        _now,
        _level,
        _src,
        msg
      ] = formatArgs;

      let cond = _.startsWith(msg, loggerCallArgs[0]);

      if (!cond) {
        // eslint-disable-next-line no-console
        console.error({
          consoleFun,
          format,
          formatArgs
        });
      }
      expect(cond).toBe(true);
    });

    await logger.error(...loggerCallArgs);
    expect(spyFormat).toHaveBeenCalledTimes(1);
  });
});
