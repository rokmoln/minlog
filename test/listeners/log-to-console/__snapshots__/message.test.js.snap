// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logToConsole listener should print a message 1`] = `
Object {
  "consoleFun": "error",
  "format": "%s %s %s
%s",
  "formatArgs": Array [
    "1970-01-01T00:00:00.000Z",
    "ERROR",
    "",
    "This is a message.",
  ],
}
`;
